### PR TITLE
Generic aggregate function state

### LINKED ISSUE #1189: Reduce the memory size of AggregateFunction state
**Summary**

This test code proves that DataValue has memory size 64 bytes, because DataValue is an enum struct which uses the maximum memory for each enum.

```
 println!("memory size -> {:?}", std::mem::size_of::<DataValue>());   // 64 
 println!("memory size -> {:?}", std::mem::size_of::<Option<u64>>()); // 16
```

We should introduce a generic state implementation  to improve memory and efficiency.

### LINKED ISSUE #1215: common/io crate to have bianary serde
**Summary**

Currently, we serde the aggregate function state by *serde/json*, but it's not efficient as binary serde.

Eg:
avg_state
Current each state item has useless keys, and it's not compact format.
```
{"sum":xxx, "count":  xxx}
```

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

- Use Generic state representation instead of DataValue
- Merge min and max  functions into the same file.
- Introduce common/io crate, now we have BinarySer and BinaryDe for state serde.

## Changelog

- Improvement
- Performance Improvement

## Related Issues

Fixes  #1189
Fixes #1215

Prepare of #966 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Review by PsiACE — COMMENTED** [review]
Good enough for me.

**Review by jyizheng — COMMENTED** [review]
It looks good to me!

File: common/io/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "common-io"
```
Comment (jyizheng) [inline common/io/Cargo.toml:2]: What are the users of this package?

Comment (sundy-li) [reply common/io/Cargo.toml:2]: see #1215

**Review by BohuTANG — APPROVED** [review]
LGTM
Thank you @sundy-li

**Pull request merged**